frontend/annotation_resolver.py

```diff
@@ -60,82 +35,138 @@ def resolve(self, annotation, solver):
+            
+            if annotation_val == "Callable":
+                # Parse Callable type
+                assert isinstance(annotation.slice.value, ast.Tuple)
+                assert len(annotation.slice.value.elts) == 2
+                assert isinstance(annotation.slice.value.elts[0], ast.List)
+
+                # Get the args and return types
+                args_annotations = annotation.slice.value.elts[0].elts
+                args_types = [self.resolve(x, solver, generics_map) for x in args_annotations]
+                return_type = self.resolve(annotation.slice.value.elts[1], solver, generics_map)
+
+                return self.z3_types.funcs[len(args_types)](*(args_types + [return_type]))
+
+            if annotation_val == "Union":
+                # Parse Union type
+                assert isinstance(annotation.slice.value, (ast.Name, ast.Tuple))
+
+                # Get the types of the union args
+                if isinstance(annotation.slice.value, ast.Name):
+                    union_args_types = [self.resolve(annotation.slice.value, solver, generics_map)]
+                else:
+                    union_args_types = [self.resolve(x, solver, generics_map) for x in annotation.slice.value.elts]
+
+                # The result of the union type is only one of args, Z3 picks the appropriate one
+                # according to the added constraints.
+                # Therefore, using more than one type from the union in the same program isn't yet supported.
+                # For example, the following is not supported:
+                # def f(x: Union[int, str]): ...
+                # f(1)
+                # f("str")
+                result_type = solver.new_z3_const("union")
+                solver.add(Or([result_type == arg for arg in union_args_types]),
+                           fail_message="Union in type annotation")
+
+                return result_type
+
+        raise ValueError("Invalid type annotation in line {}".format(annotation.lineno))
+
+    def add_annotated_function_axioms(self, args_types, solver, annotations, result_type):
+        """Add axioms for a function call to an annotated function
+        
+        Reprocess the type annotations for every function call to prevent binding a certain type
+        to the function definition
+        """
+        args_annotations = annotations[0]
+        result_annotation = annotations[1]
+
+        if len(args_types) != len(args_annotations):
+            raise TypeError("The function expects {} arguments. {} were given.".format(len(args_annotations),
+                                                                                       len(args_types)))
+        generics_map = {}
+
+        for i in range(len(args_annotations)):
+            arg_type = self.resolve(args_annotations[i], solver, generics_map)
+            solver.add(solver.z3_types.subtype(args_types[i], arg_type),
+                       fail_message="Generic parameter type")
```

Comment:
If we wanted to do that, I think there should just be another (optional) parameter where you could pass the AST node, and then the line information or pretty-printed node would automatically be attached to the error message.